pkg/kv/kvserver/apply/asserter.go

```diff
@@ -0,0 +1,308 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package apply
+
+import (
+	fmt "fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverbase"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"go.etcd.io/raft/v3/raftpb"
+)
+
+// Asserter is a test utility that tracks application of Raft commands, and
+// primarily asserts that:
+//
+//   - A request is only applied once, at a single log index and lease applied index
+//     across all replica (i.e. no double-applies or replays).
+//
+//   - A replica's applied index, applied term, lease applied index, and closed timestamp
+//     do not regress (assuming no node restarts).
+type Asserter struct {
+	mu syncutil.Mutex
+	// proposedCmds tracks proposed commands by ID, along with the proposer.
+	// Reproposals under new command IDs are considered separate proposals here.
+	proposedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]roachpb.ReplicaID
+	// reproposedCmds keeps track of LAI reproposals under different command IDs
+	// (when the LAI changes), keyed by the original seed proposal's command ID
+	// and the reproposal's command ID.
+	reproposedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]map[kvserverbase.CmdIDKey]bool
+	// seedCmds maps a LAI reproposal command ID back to its original seed proposal.
+	// It is the reverse index of reproposedCmds.
+	seedCmds map[roachpb.RangeID]map[kvserverbase.CmdIDKey]kvserverbase.CmdIDKey
```

Comment:
Is this level of detail needed here?

The contract that we're trying to verify is that each command is applied at most once.

Suppose we have a canonical command ID (which we do: it's `seedProposal`'s command ID; or the command's own ID if it is the seed). Then we just need to make sure that this ID is mentioned only once at exactly one log index etc.

Reproposals are an implementation detail that's not part of the visible behaviour. A command can be proposed, committed, and [almost] applied many times, but ultimately will apply only once.

```
command: proposal(LAI_1) -> reproposal(LAI_2) -> ... reproposal(LAI_k)
             |                    |                    |
             v                    v                    v
      reproposal(LAI_1)_1   reproposal(LAI_2)_1  ...  ...
             |                    |                    |
             v                    v                    v
      reproposal(LAI_1)_2   reproposal(LAI_2)_2.     reproposal(LAI_k)_m
          ...

             ^                    ^                    ^
             |                    |                    |
      all these are either never committed or     only one of these reproposals applies
      apply as no-ops (so Apply is not called)
```

If the apply contract works, the `Apply` function will ever be called only on one of these reproposals. If it's called on another one, this is a violation so we stop immediately.

Thus, it seems like we don't need to track the reproposals concept in the first place? Just the first occurrence.

Reply:
I think this is a balance. One inclination is to check as many invariants as possible. Another extreme is check only the things that this harness makes possible (which were impossible by other means): "global" invariants / safety properties. A bunch of per-replica/per-proposal properties can probably be verified inline as runtime assertions. This harness should mainly focus on what's not trivial or not possible locally.

---

One problem with the `seedProposal` though is that it's populated only on the leaseholder. Followers don't know if proposals with different IDs are related. This means that when the leaseholder `Apply` a reproposal, we will see a non-empty `seedProposal` etc; when a follower applies the same reproposal, we won't see it.


A snapshot applied on a leaseholder can render in skipping a few leaseholder-side `Apply` calls that carry the `seedProposal` signal. I.e. the ordering [follower-side `Apply` -> snapshot at leaseholder that covers this applied entry -> missing leaseholder-side `Apply`] introduces a gap in the double-apply detection.

To have 100% coverage in double-apply detection, we should either:
- (in this PR) intercept proposals on the leaseholder at stages earlier than apply, to memoize the ID->seedID mapping; then use this mapping to canonicalize the command IDs that come through the `Apply` call
- or (preferrably, longer-term) go back to having a command ID that is properly reused across reproposals.

Reply:
>One problem with the seedProposal though is that it's populated only on the leaseholder. Followers don't know if proposals with different IDs are related. This means that when the leaseholder Apply a reproposal, we will see a non-empty seedProposal etc; when a follower applies the same reproposal, we won't see it.

We record the seed/reproposal mapping during `Propose` (where we always know the seed), and recover it in `Apply`  regardless of which replica it's applying on.

>A bunch of per-replica/per-proposal properties can probably be verified inline as runtime assertions. This harness should mainly focus on what's not trivial or not possible locally.

Yes, we should add runtime assertions in the actual code as well, and that should be preferred. For now, I'm just trying to get some basic coverage in for #114421. You should feel free to make whatever changes you want later.

Reply:
Nice, having a `Propose`+`Apply` pair of trackers (for this and other reasons) makes sense.

Reply:
Rewrote the logic to track the command ID a seed command eventually applies as.